torch/csrc/Device.cpp

```diff
@@ -0,0 +1,217 @@
+  oss << "Device(device_type=\'" << deviceTypeString(self->device_type) << "\'";
+  if (!self->is_default) {
+    oss << ", device_index=" << self->device_index;
+  }
+  oss << ")";
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_str(THPDevice*self)
+{
+  std::ostringstream oss;
+  if (!self->is_default) {
+    oss << deviceTypeString(self->device_type) << ":" << self->device_index;
+  } else {
+    oss << deviceTypeString(self->device_type);
+  }
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+    "Device(Device device)",
+    "Device(String device_type, int64_t? device_index=-1)"
+  });
+  torch::ParsedArgs<2> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  if (r.idx == 0) {
+    auto device = r.device(0);
+    return THPDevice_New(device.device_type, device.device_index, device.is_default);
+  } else if (r.idx == 1) {
+    auto as_device = r.device(0);  // this works, because device can take strings
+    auto device_type = r.string(0);
+    if (!as_device.is_default) {
+      throw std::runtime_error("device_type must not include index because index is passed explicitly " + device_type);
+    }
+
+    auto is_default = r.isNone(1);
+    auto device_index = r.toInt64WithDefault(1, -1);
+    // make sure this is constructible
+    auto device = torch::Device(as_device.device_type, device_index, is_default);
+    return THPDevice_New(device.device_type, device.device_index, device.is_default);
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_type(THPDevice *self)
+{
+  HANDLE_TH_ERRORS
+  return THPUtils_packString(deviceTypeString(self->device_type));
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_index(THPDevice *self)
+{
+  HANDLE_TH_ERRORS
+  if (self->is_default) {
+    Py_RETURN_NONE;
+  } else {
+    return THPUtils_packInt64(self->device_index);
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject *THPDevice_cuda_index(THPDevice *self)
+{
+  HANDLE_TH_ERRORS
+  if (self->device_type == torch::DeviceType::CUDA) {
+    return THPUtils_packInt64(self->device_index);
+  }
+  std::ostringstream oss;
+  oss << "cuda_index only supported on cuda device, got: " << deviceTypeString(self->device_type);
+  throw std::runtime_error(oss.str());
+  END_HANDLE_TH_ERRORS
+}
+
+static bool THPDevice_equal(THPDevice *a, THPDevice *b) {
+  return a->device_type == b->device_type && a->device_index == b->device_index && a->is_default == b->is_default;
+}
+
+PyObject *THPDevice_rc(PyObject *a, PyObject *b, int op) {
+  HANDLE_TH_ERRORS
+  if (!THPDevice_Check(a)) {
+    throw torch::TypeError("Device can only be compared with DeviceType, got %s", Py_TYPE(a)->tp_name);
+  }
+  if (!THPDevice_Check(b)) {
+    throw torch::TypeError("Device can only be compared with DeviceType, got %s", Py_TYPE(b)->tp_name);
```

Comment:
In Python one usually does `return NotImplemented`, and lets Python do the error handling (or a dispatch to `__r*__` method of `b`). Is there an equivalent here?

Reply:
You are right, these should be Py_NotImplemented; TypeError is used for comparison types that aren't implemented.

Reply:
According to: https://docs.python.org/2/c-api/typeobj.html#c.PyTypeObject.tp_richcompare